envoy/router/router.h

```diff
@@ -310,6 +313,11 @@ class RetryPolicy {
    * back-off interval parsed from response headers.
    */
   virtual std::chrono::milliseconds resetMaxInterval() const PURE;
+
+  /**
+   * @return RetryPolicyConstSharedPtr a shared pointer to this instance.
+   */
+  RetryPolicyConstSharedPtr getConstShared() const { return shared_from_this(); }
```

Comment:
It might not work for null route entry and is also a hacky way to turn a reference into a shared pointer. Can you add a method to the RouteEntry interface to also return a shared pointer in addition to reference? I think it will only a change to the Null route impl. We can mark the reference returning method as deprecated and remove it separately as it is a very big code churn. What do you think?

What do you think?